### DISCUSSION TITLE
List of (soft) recommendations or "best practices" for implementations

### DISCUSSION THREAD

**Discussion by jgonggrijp** [mustache/spec#164] [Ideas]
The spec describes what an implementation should do in order to ensure that templates are portable. Besides that, there could also exist a more informal and optional list of guidelines, which describe how an implementation could make itself convenient for users. Credits to @bobthecow for suggesting this in https://github.com/mustache/spec/discussions/156#discussioncomment-7493815.

Right now, I can think of three candidates for such a list:

- ambient context (#156);
- liberated lambdas (https://github.com/mustache/spec/discussions/155#discussioncomment-7515586);
- pragma tags (I've seen these mentioned in past comments by @bobthecow and @groue).

If we were to make such a list, I think it should be "free for all", i.e., not subject to review or consensus. Partly because implementers are free to ignore the recommendations anyway, partly because we must not burden ourselves with new processes that need to be controlled. Everyone can add and nobody should remove.

I think the wiki might be a suitable place for such a list.

**Comment by agentgt** [comment]
I firmly believe a best practice is providing some data variable that is a boolean of `first` and `last` while in a list section.

Whether that is `@first` or `@last` or `-first` or `-last` I think doesn't really matter when it comes to a "best practice".

That is most Mustache implementations should provide a means to create a comma separated list without inducing an aneurism of having to dig in various complicated implementation tricks.

In traditional HTML where you also have Javascript and CSS at play knowing first and last was not as important.

Now that React and various Javascript frameworks have taken over a good portion of web site rendering Mustache's target output media I think is becoming more mixed and varied than just HTML. 

For example I use Mustache a great deal for Java/SQL code generation and code generation in so many cases requires generating a list separated by commas in them where a trailing `,` is not allowed (for example SQL).

Every time I reach for JMustache or now JStachio (both which have virtual keys) for this problem because it is not obvious how to do this instantly with Mustache.java (the other reason is JMustache/JStachio has more consistent whitespace).

**Reply by jgonggrijp** [reply]
Does it really need to be a boolean data variable? What if I provide a small library of lamdas and filters, one them being an `enumerate` that unfolds each list element into a `index`, `value`, `first`, `last` quartet if the user opts in? Would I still be following your recommendation in that case?

**Reply by bobthecow** [reply]
my take is that best practices are for stuff that could be done in any number of ways, but that are often done a specific way by implementations. they might be language dependent, or open to interpretation based on idiomatic usage or pre-existing patterns. this specific case feels like a "get it in the spec" thing to me.

**Reply by bobthecow** [reply]
though i'm with @jgonggrijp. i'd personally use a filter for this, and possibly supply it as a "helper" in the implementation?

**Reply by bobthecow** [reply]
Concretely, for Mustache.php it looks something like this:

```php
<?php

class EnumeratedValue {
    private $index;
    private $value;
    private $first;
    private $last;

    private const KNOWN_KEYS = ['@index', '@key', '@value', '@first', '@last', '@only'];

    public function __construct($index, $value, $first, $last) {
        $this->index = $index;
        $this->value = $value;
        $this->first = $first;
        $this->last = $last;
    }

    /**
     * This whole thing for a __toString so we can support {{.}} :)
     */
    public function __toString() {
      return (string) $this->value;
    }

    public function __get($key) {
      switch ($key) {
        case '@index':
        case '@key':
          return $this->index;
        case '@value':
          return $this->value;
        case '@first':
          return $this->first;
        case '@last':
          return $this->last;
        case '@only':
          return $this->first && $this->last;
      }
    }

    public function __isset($key) {
      return in_array($key, self::KNOWN_KEYS);
    }
}

$m = new Mustache_Engine([
  'helpers' => [
    'enumerate' => function($arr) {
      // TODO: enforce that this is enumerable?

      $first = array_key_first($arr);
      $last = array_key_last($arr);

      foreach ($arr as $k => $v) {
        yield new EnumeratedValue($k, $v, $k === $first, $k === $last);
      }
    },
  ]
]);

$tpl = <<<EOS
{{%FILTERS}}
{{#values | enumerate}}
  {{@index}}: {{.}}{{^ @last }},{{/ @last }}
{{/values}}
EOS;

echo $m->render($tpl, [
  'values' => [
    'a',
    'b',
    'c',
  ],
]);

//  0: a,
//  1: b,
//  2: c
```

**Reply by agentgt** [reply]
@jgonggrijp 
> Does it really need to be a boolean data variable?

No but that is the easiest, how most other templating languages do it and possibly fastest (although that reason is far less important).

For example the filter way would require object allocation instead of stack allocation in the case of Java.

> What if I provide a small library of lamdas and filters, one them being an enumerate that unfolds each list element into a `index`, `value`, `first`, `last` quartet if the user opts in? Would I still be following your recommendation in that case?

Sure if its provided OOB hence the best practice part. e.g. provide away OOB to do this basic thing that every other templating language can do.

I will say I think:

>  `index`, `value`, `first`, `last`

will have much greater chance of collision across implementations than some special prefix unless masking is fixed but that is probably a separate topic. Besides your filter I guess could follow a more unique name... 

I also suppose to avoid performance issues implementations could do special things for builtin intrinsic filters.


@bobthecow 
>  this specific case feels like a "get it in the spec" thing to me.

Ideally yes but I'm willing to make concessions for those cases where a filter is better for some implementations or for those that want to remain pure for whatever reasons.


BTW this has real world repercussions because I believe the openapi code generation guys dumped Mustache for handlebars.java in large part because of this problem.

In fact one of them is now using JStachio: https://github.com/jskov/openapi-jaxrs-client

Many of their complex templates unsurprisingly is using some form of `first`,`last`: https://github.com/jskov/openapi-jaxrs-client/blob/26eedea4ca07c789252eaba563ca84099219474b/modules/generator/src/main/resources/templates/api.mustache#L73

**Reply by agentgt** [reply]
To put it simpler my thoughts on filters and some mechanism to achieve or first/last do not have to be at odds. I do not care if it is virtual keys or filters that solve the problem and I doubt most users do. 

What they do not want to do is go figure out how wire in that `EnumeratedValue` into the mustache engine. Even though it is shown to be easy with Mustache PHP it is not so obvious or easy in other languages/implementations. The best practice part is implementers should make that possible, **documented** (in the case of Mustache.java) and/or on by default.

Also filters are inherently more complicated. You need to first find the filter (and or register) and then figure out what it returns and then make sure it call its bindings all of which is implementation specific.

Then you have... well should explain what filters are or what is happening with `{{#list | enumerate}}`.

I think most people can safely describe first/last/index in less than a paragraph. I don't think filters as they are currently being proposed can be described so easily: https://github.com/mustache/spec/discussions/153#discussion-5812473

Perhaps the above will be remedied over time with simpler description but currently it is a lot more than just this key that is available while in a list.

Then there is implementation questions revolving filters that I can go add to #153 like how much `|` short circuited?

I'm definitely a fan of filters and can probably get a single `|` (pipe) working but multiple piping has some evaluation order (e.g. lazy or eager) and static type checking that I may not be able to easily resolve.

**Reply by jgonggrijp** [reply]
I agree with @agentgt that trailing commas are currently needlessly difficult in Mustache, especially if you want to stay portable (because then the only option is to decorate the data before passing them to the template).

I agree with @bobthecow that a recommendation to have "some kind of solution" for this will not really address the problem. In fact, it will make templates less portable, because users will rely on a feature that one implementation provides but the other does not.

For example, suppose that Wontache implements virtual keys that go `-index`, `-this`, `-first`, `-last` and Mustache.php implements virtual keys that go `@index`, `@current`, `@first`, `@last`. We both follow the recommendations, but users cannot bring their templates from Wontache to Mustache.php without editing them.

To really address the problem, we would either specify the exact virtual keywords and their meanings, or we would specify some other, more general feature that can be used to solve trailing commas.

By the way, at this point I think that the trailing comma problem should probably have a discussion of its own.

If, for the sake of argument, we assume that we specify filters, then there is some room for a soft recommendation again: to ship with a ready to use `enumerate` filter that can do first/last detection. Even if Mustache.php follows that recommendation but Wontache doesn't (or does but with slightly different semantics), users will still be able to reuse their templates. All they need to do, is port the Mustache.php version of `enumerate` to Wontache.

You express valid concerns with regard to the explainability and implementability of filters; I will reply to those in new comments in #153.

**Reply by agentgt** [reply]
I agree largely what you wrote however:

> For example, suppose that Wontache implements virtual keys that go -index, -this, -first, -last and Mustache.php implements virtual keys that go @index, @current, @first, @last. 

While perhaps not instantly portable it is much easier to port than a filter that does it. In fact in some places where we replaced JMustache with Handlebars we just did a simple to search and replace of (`-first` -> `@first`).

One con about the filter on porting is that decorating object returned by the enumerator may not do the correct thing for: `{{.}}`. That is a naive `toString` in @bobthecow implementation may not work or even be possible in some languages.

That is in some languages it is hard to make generic like functions that take any like objects. OCaml comes to mind.

In my implementation it is a big deal. The compiler will block you from trying to `toString` arbitrary objects.

That is assume `EnumeratedValue` is the current node.

```
{{.}}
```

Would fail. Not even at runtime but compile time. You can read about that more here: https://jstach.io/jstachio/io.jstach.jstache/io/jstach/jstache/JStacheFormatterTypes.html

So you would have to register `EnumeratedValue` as safe `toString` but how does `EnumeratedValue` know the contained object is safe `toString`?

One of my favorite things about Mustache is how safe it can be made to be.

**Reply by jgonggrijp** [reply]
I see what you mean. `{{.}}` and direct lookups like `{{name}}` instead of `{{@value.name}}` are problematic for portability. I kind of skipped over that the first time I read @bobthecow's code (he doesn't do the direct lookups but I mention it because Handlebars does support that).

I could implement those shortcuts in Wontache, but I honestly don't want to. I was planning to leave off the `@` markers and just force users to write `{{value}}` and `{{value.name}}` (or `{{#value}}...{{name}}...{{/value}}`). It makes the implementation much simpler and the semantics within the template more straightforward.

If it comes to a soft recommendation for an `enumerate`-like filter, I suggest including the following. Implementations that support shortcuts like `{{.}}`, should clearly state in the documentation that this notation may not be portable. Users can then make an informed choice between notational convenience and portability, and implementers can provide necessary functionality without having to jump hoops.

**Reply by agentgt** [reply]
@bobthecow @jgonggrijp  I showed a JStachio implementation of enumerate here: https://github.com/mustache/spec/discussions/153#discussioncomment-7527279

It respects the original `{{.}}` however it is not a filter as currently defined.

**Comment by jgonggrijp** [comment]
#174 might be a subject for this list as well, although it probably deserves a hard rule in a 2.0 version of the spec.
